st2common/tests/unit/test_rbac_resolvers_rule_enforcement.py

```diff
@@ -0,0 +1,427 @@
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role with no permission grants, should return false for everything
+        user_db = self.users['1_custom_role_no_permissions']
+        self.assertFalse(self._user_has_resource_db_permissions(
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role with unrelated permission grant to parent pack
+        user_db = self.users['custom_role_pack_grant']
+        self.assertFalse(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_1'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+        self.assertFalse(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_2'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+
+        # Custom role with with grant on the parent pack
+        user_db = self.users['custom_role_rule_pack_grant']
+        self.assertTrue(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_1'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+        self.assertTrue(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_2'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+
+        # Custom role with a direct grant on rule
+        user_db = self.users['custom_role_rule_grant']
+        self.assertTrue(resolver.user_has_resource_db_permission(
+            user_db=user_db,
+            resource_db=self.resources['rule_enforcement_3'],
+            permission_type=PermissionType.RULE_ENFORCEMENT_VIEW))
+
+        # Custom role - "rule_all" grant on the rule parent pack
+        user_db = self.users['custom_role_pack_rule_all_grant']
+        resource_db = self.resources['rule_enforcement_1']
+        self.assertTrue(self._user_has_resource_db_permissions(
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role - "rule_all" grant on the rule
+        user_db = self.users['custom_role_rule_all_grant']
+        resource_db = self.resources['rule_enforcement_1']
+        self.assertTrue(self._user_has_resource_db_permissions(
+            resolver=resolver,
+            user_db=user_db,
+            resource_db=resource_db,
+            permission_types=all_permission_types))
+
+        # Custom role - "rule_modify" grant on rule_1
+        # XXX: This test is failing because I think there is a bug in
```

Comment:
It's worth noting that those are "implicit" grants meaning we don't write anything to database or file system (implicit grant resolving happens inside the resolver).

See https://github.com/StackStorm/st2/blob/master/st2common/st2common/rbac/resolvers.py#L216 for an example